src/message_passing.cpp

```diff
@@ -17,6 +17,29 @@ extern "C" bool openmc_master()
   return mpi::master;
 }
 
+vector<int64_t> calculate_parallel_index_vector(const int64_t size)
+{
+  vector<int64_t> result;
+  result.reserve(n_procs + 1);
+
+#ifdef OPENMC_MPI
+  result.resize(n_procs);
+  vector<int64_t> bank_size(n_procs);
+
+  // Populate the result with cumulative sum of the number of
+  // surface source banks per process
+  MPI_Scan(&size, bank_size.data(), 1, MPI_INT64_T, MPI_SUM, intracomm);
+  MPI_Allgather(
+    bank_size.data(), 1, MPI_INT64_T, result.data(), 1, MPI_INT64_T, intracomm);
+  result.insert(result.begin(), 0);
```

Comment:
Inserting at position zero requires reallocating -- I think it'd be cleaner to `resize(n_procs + 1)` above and then `Allgather` into `result.data() + 1`.

Reply:
Ah sorry, just realizing this is not code that _you_ wrote and was just moved from elsewhere. Feel free to disregard, or update if you feel so inclined!